### PR TITLE
[MacOS] On Mac, use the copyfile API to copy files.

### PR DESCRIPTION
**Pull request by migueldeicaza**
Copyfile is a high level API that does all the heavy lifting for
copying files and does this in a Mac-correct way.   We configure it
to copy both the contents as well as the metadata and extended attributes
in the file.

Additionally, this uses COPYFILE_CLONE which will attempt to use the
APFS clone capabilities for fast copies.

On my Mac with APFS, copying a 5 gig file (Xcode) goes from taking
7.2 seconds to copy on an SSD to 0.5 seconds.

Fixes: https://github.com/mono/mono/issues/9030

### CODE REVIEW

File: mono/metadata/w32file-unix.c [hunk 1]
```diff
@@ -14,6 +14,12 @@
 #ifdef HAVE_SYS_STATVFS_H
 #include <sys/statvfs.h>
 #endif
+#ifdef HAVE_COPYFILE_H
+#include <copyfile.h>
+#  if !defined(COPYFILE_CLONE)
```
Comment (luhenry) [inline mono/metadata/w32file-unix.c]: `#ifndef` and no indentation, for consistency with rest of file.

File: mono/metadata/w32file-unix.c [hunk 2]
```diff
@@ -2359,7 +2365,17 @@ CopyFile (const gunichar2 *name, const gunichar2 *dest_name, gboolean fail_if_ex
 		
 		return(FALSE);
 	}
-	
+
+#if HAVE_COPYFILE_H
```
Comment (luhenry) [inline mono/metadata/w32file-unix.c]: `#ifdef` to be consistent with https://github.com/mono/mono/pull/9038/files#diff-e3902264b6c6cb537e9eec9149668e83R17

File: mono/metadata/w32file-unix.c [hunk 3]
```diff
@@ -14,6 +14,12 @@
 #ifdef HAVE_SYS_STATVFS_H
 #include <sys/statvfs.h>
 #endif
+#ifdef HAVE_COPYFILE_H
+#include <copyfile.h>
+#  if !defined(COPYFILE_CLONE)
+#    #define COPYFILE_CLONE 0
```
Comment (lewurm) [inline mono/metadata/w32file-unix.c]: why is this needed?

File: mono/metadata/w32file-unix.c (hunk 1 above)
Comment (migueldeicaza) [reply mono/metadata/w32file-unix.c]: No - that is a bad style that someone went and added in recent years.   Undo that disaster please.

File: mono/metadata/w32file-unix.c (hunk 3 above)
Comment (migueldeicaza) [reply mono/metadata/w32file-unix.c]: Older MacOS versions do not have COPYFILE_CLONE

File: mono/metadata/w32file-unix.c (hunk 2 above)
Comment (migueldeicaza) [reply mono/metadata/w32file-unix.c]: The top of the file uses #ifdef, the rest of the file uses #if.

Wont make this change.

File: mono/metadata/w32file-unix.c (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/w32file-unix.c]: Can you unconditionally define it to the correct value?
Or you want 0 (incorrect value?) as no-op on old systems?

Comment (jaykrell) [reply mono/metadata/w32file-unix.c]: Correct value is #define COPYFILE_CLONE          (1<<24)
but intent might be to make it a no-op downlevel.

**Comment by luhenry** [comment]
@migueldeicaza this seems to break quite a few tests on macOS:
 - https://jenkins.mono-project.com/job/test-mono-pull-request-i386-osx/11085/testReport/
 - https://jenkins.mono-project.com/job/test-mono-pull-request-amd64-osx/11092/testReport/

File: mono/metadata/w32file-unix.c (hunk 3 above)
Comment (migueldeicaza) [reply mono/metadata/w32file-unix.c]: Good idea

**Comment by migueldeicaza** [comment]
Submitting a new patch, to prevent the breakage in IsolatedStorage, the subtle semantics of Copy and should fix the MSBuild issue.

File: mono/metadata/w32file-unix.c [hunk 4]
```diff
@@ -2359,7 +2365,17 @@ CopyFile (const gunichar2 *name, const gunichar2 *dest_name, gboolean fail_if_ex
 		
 		return(FALSE);
 	}
-	
+
+#if HAVE_COPYFILE_H
+	ret = copyfile (utf8_src, utf8_dest, NULL, COPYFILE_ALL | COPYFILE_CLONE | (fail_if_exists ? COPYFILE_EXCL : 0));
```
Comment (filipnavara) [inline mono/metadata/w32file-unix.c]: This won't work because COPYFILE_CLONE implies COPYFILE_EXCL (https://opensource.apple.com/source/copyfile/copyfile-146.30.2/copyfile.c.auto.html). One way to solve that is to fallback on `if (ret != 0 && errno == EEXIST && ! fail_if_exists)` to non-cloning variant. I wanted to check why COPYFILE_UNLINK didn't work, but the build results are gone now :-/

Comment (filipnavara) [reply mono/metadata/w32file-unix.c]: Nvm, I see it now - same file name for src and dest is broken for COPYFILE_UNLINK.
